src/analysis/calculus/implicit.lean

```diff
@@ -0,0 +1,308 @@
+  φ.pt ∈ φ.to_local_homeomorph.source :=
+φ.has_strict_fderiv_at.mem_to_local_homeomorph_source
+
+lemma map_pt_mem_to_local_homeomorph_target :
+  (φ.left_fun φ.pt, φ.right_fun φ.pt) ∈ φ.to_local_homeomorph.target :=
+φ.to_local_homeomorph.map_source $ φ.pt_mem_to_local_homeomorph_source
+
+lemma prod_map_implicit_function :
+  ∀ᶠ (p : F × G) in 𝓝 (φ.prod_fun φ.pt), φ.prod_fun (φ.implicit_function p.1 p.2) = p :=
+φ.has_strict_fderiv_at.eventually_right_inverse.mono $ λ ⟨z, y⟩ h, h
+
+lemma left_map_implicit_function :
+  ∀ᶠ (p : F × G) in 𝓝 (φ.prod_fun φ.pt), φ.left_fun (φ.implicit_function p.1 p.2) = p.1 :=
+φ.prod_map_implicit_function.mono $ λ z, congr_arg prod.fst
+
+lemma right_map_implicit_function :
+  ∀ᶠ (p : F × G) in 𝓝 (φ.prod_fun φ.pt), φ.right_fun (φ.implicit_function p.1 p.2) = p.2 :=
+φ.prod_map_implicit_function.mono $ λ z, congr_arg prod.snd
+
+lemma implicit_function_apply_image :
+  ∀ᶠ x in 𝓝 φ.pt, φ.implicit_function (φ.left_fun x) (φ.right_fun x) = x :=
+φ.has_strict_fderiv_at.eventually_left_inverse
+
+lemma to_implicit_function
+  (g'inv : G →L[𝕜] E) (hg'inv : φ.right_deriv.comp g'inv = continuous_linear_map.id 𝕜 G)
+  (hg'invf : φ.left_deriv.comp g'inv = 0) :
+  has_strict_fderiv_at (φ.implicit_function (φ.left_fun φ.pt)) g'inv (φ.right_fun φ.pt) :=
+begin
+  have := φ.has_strict_fderiv_at.to_local_inverse,
+  simp only [prod_fun] at this,
+  convert this.comp (φ.right_fun φ.pt)
+    ((has_strict_fderiv_at_const _ _).prod (has_strict_fderiv_at_id _)),
+  simp only [continuous_linear_map.ext_iff, continuous_linear_map.coe_comp', function.comp_app]
+    at hg'inv hg'invf ⊢,
+  simp [continuous_linear_equiv.eq_symm_apply, *]
+end
+
+
+end implicit_function_data
+
+namespace has_strict_fderiv_at
+
+section complemented
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+  {E : Type*} [normed_group E] [normed_space 𝕜 E] [complete_space E]
+  {F : Type*} [normed_group F] [normed_space 𝕜 F] [complete_space F]
+  {f : E → F} {f' : E →L[𝕜] F} {a : E}
+
+section defs
+
+variables (f f')
+
+/-- Data used to apply generic implicit function theorem to the case of a strictrly differentiable
+ map such that its derivative is surjective has has a complemented kernel. -/
+@[simp] def implicit_function_data_of_complemented (hf : has_strict_fderiv_at f f' a)
+  (hf' : f'.range = ⊤) (hker : f'.ker.closed_complemented) :
+  implicit_function_data 𝕜 E F f'.ker :=
+{ left_fun := f,
+  left_deriv := f',
+  right_fun := λ x, classical.some hker (x - a),
+  right_deriv := classical.some hker,
+  pt := a,
+  left_has_deriv := hf,
+  right_has_deriv := (classical.some hker).has_strict_fderiv_at.comp a
+    ((has_strict_fderiv_at_id a).sub_const a),
+  left_range := hf',
+  right_range := linear_map.range_eq_of_proj (classical.some_spec hker),
+  is_compl_ker := linear_map.is_compl_of_proj (classical.some_spec hker) }
+
+/-- A local homeomorphism between E` and `F × f'.ker` sending level surfaces of `f`
+to horizontal subspaces. -/
```

Comment:
Should it be vertical subspaces? Your model for `f` is the first projection, with kernel the vertical space `f.ker = f'.ker`.

Reply:
```suggestion
to vertical subspaces. -/
```